### PR TITLE
Adds and implements node taints count metric.

### PR DESCRIPTION
**Pull request by piotrwrotniak**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Adds and implements node taints count metric.

#### Does this PR introduce a user-facing change?
```release-note
Adds and implements node taints count metric.
```

### CODE REVIEW

File: cluster-autoscaler/utils/taints/taints.go [hunk 1]
```diff
@@ -416,3 +442,50 @@ func FilterOutNodesWithStartupTaints(taintConfig TaintConfig, allNodes, readyNod
...
 	return newAllNodes, newReadyNodes
 }
+
+// CountNodeTaints counts used node taints.
+func CountNodeTaints(nodes []*apiv1.Node, taintConfig TaintConfig) map[string]int {
+	foundTaintsCount := make(map[string]int)
+	for _, node := range nodes {
+		for _, taint := range node.Spec.Taints {
+			key := getTaintTypeToReport(taint.Key, taintConfig)
+			if _, ok := foundTaintsCount[key]; ok {
```
Comment (x13n) [inline cluster-autoscaler/utils/taints/taints.go:452]: Missing map entries will default to 0, so there's no need to handle "not found" case separately - simple increment will suffice.

File: cluster-autoscaler/utils/taints/taints.go [hunk 2]
```diff
@@ -108,6 +117,23 @@ var (
 	// Mutable only in unit tests
 	maxRetryDeadline      time.Duration = 5 * time.Second
 	conflictRetryInterval time.Duration = 750 * time.Millisecond
+
+	explicitlyReportedNodeTaints = TaintKeySet{
```
Comment (x13n) [inline cluster-autoscaler/utils/taints/taints.go:121]: This is duplicating the list from `NodeConditionTaints`, can we initialize this with `NodeConditionTaints` with a few extra taints? I worry they may get out of sync.

File: cluster-autoscaler/proposals/metrics.md [hunk 3]
```diff
@@ -21,17 +21,18 @@ All the metrics are prefixed with `cluster_autoscaler_`.
...
+| max_nodes_count | Gauge |                                            | Maximum number of nodes in all node groups.                                        |
+| cluster_cpu_current_cores | Gauge |                                            |                                                                                    | Current number of cores in the cluster, minus deleting nodes. |
+| cpu_limits_cores | Gauge | `direction`=&lt;`minimum` or `maximum`&gt; | Minimum and maximum number of cores in the cluster.                                |
+| cluster_memory_current_bytes | Gauge |                                            | Current number of bytes of memory in the cluster, minus deleting nodes.            |
+| memory_limits_bytes | Gauge | `direction`=&lt;`minimum` or `maximum`&gt; | Minimum and maximum number of bytes of memory in cluster.                          |
+ | node_taints_count | Gauge | `type`=&lt;taint-type&gt;                  | Number of taints per type in cluster.                                              |
```
Comment (x13n) [inline cluster-autoscaler/proposals/metrics.md:35]: This is a pretty old proposal, not sure if there's much value in adding the metric here. There's already a lot of metrics that were added in the code but not to this doc.

File: cluster-autoscaler/utils/taints/taints.go [hunk 4]
```diff
@@ -416,3 +442,50 @@ func FilterOutNodesWithStartupTaints(taintConfig TaintConfig, allNodes, readyNod
...
+	// Track deprecated taints.
+	if strings.HasPrefix(key, IgnoreTaintPrefix) {
+		return IgnoreTaintPrefix
+	}
+
+	if _, ok := explicitlyReportedNodeTaints[key]; ok {
+		return key
+	}
+
+	if _, ok := taintConfig.StartupTaints[key]; ok {
```
Comment (x13n) [inline cluster-autoscaler/utils/taints/taints.go:472]: optional: There's already 3-4 places in this file that separately check prefix and exact values in `taintConfig`. WDYT about slightly refactoring this to make fields in `TaintConfig` private and expose `IsStartupTaint`/`IsStatusTaint` functions instead?

**Review by x13n — CHANGES_REQUESTED** [review]
Thanks for the change!

File: cluster-autoscaler/utils/taints/taints.go (hunk 1 above)
Comment (piotrwrotniak) [reply cluster-autoscaler/utils/taints/taints.go]: Done

File: cluster-autoscaler/utils/taints/taints.go (hunk 2 above)
Comment (piotrwrotniak) [reply cluster-autoscaler/utils/taints/taints.go]: Changed, PTAL

File: cluster-autoscaler/utils/taints/taints.go (hunk 4 above)
Comment (piotrwrotniak) [reply cluster-autoscaler/utils/taints/taints.go]: Added

File: cluster-autoscaler/proposals/metrics.md (hunk 3 above)
Comment (piotrwrotniak) [reply cluster-autoscaler/proposals/metrics.md]: Removed

**Pull request merged**